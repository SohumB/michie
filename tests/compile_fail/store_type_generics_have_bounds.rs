// This test does not test a feature. Quite the opposite.
// It proves an undesired limitation.
use core::marker::PhantomData;
use michie::memoized;

#[derive(Default)]
struct Store<K: Eq, V: Ord> {
    k: PhantomData<K>,
    v: PhantomData<V>,
}
impl<K: Eq, V: Ord> Store<K, V> {
    fn insert(&mut self, _key: K, _value: V) {}
    fn get(&self, _key: &K) -> Option<&V> {
        None
    }
}
#[memoized(key_expr = input, store_type = Store)]
fn f(input: usize) -> usize {
    input
}

fn main() {}
