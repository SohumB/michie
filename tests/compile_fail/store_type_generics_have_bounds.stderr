error[E0277]: the trait bound `K: std::cmp::Eq` is not satisfied
  --> tests/compile_fail/store_type_generics_have_bounds.rs:17:1
   |
17 | #[memoized(key_expr = input, store_type = Store)]
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `std::cmp::Eq` is not implemented for `K`
   |
note: required by a bound in `Store`
  --> tests/compile_fail/store_type_generics_have_bounds.rs:7:17
   |
7  | struct Store<K: Eq, V: Ord> {
   |                 ^^ required by this bound in `Store`
   = note: this error originates in the attribute macro `memoized` (in Nightly builds, run with -Z macro-backtrace for more info)
help: consider further restricting this bound
   |
17 | #[memoized(key_expr = input, store_type = Store)] + std::cmp::Eq
   |                                                   ++++++++++++++

error[E0277]: the trait bound `R: Ord` is not satisfied
  --> tests/compile_fail/store_type_generics_have_bounds.rs:17:1
   |
17 | #[memoized(key_expr = input, store_type = Store)]
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `Ord` is not implemented for `R`
   |
note: required by a bound in `Store`
  --> tests/compile_fail/store_type_generics_have_bounds.rs:7:24
   |
7  | struct Store<K: Eq, V: Ord> {
   |                        ^^^ required by this bound in `Store`
   = note: this error originates in the attribute macro `memoized` (in Nightly builds, run with -Z macro-backtrace for more info)
help: consider further restricting this bound
   |
17 | #[memoized(key_expr = input, store_type = Store)] + std::cmp::Ord
   |                                                   +++++++++++++++
